From d299193ea096d174fbd2841ecf07a65368b31d1f Mon Sep 17 00:00:00 2001
From: Aikar <aikar@aikar.co>
Date: Fri, 15 Feb 2013 20:10:44 -0500
Subject: [PATCH] New Old Network Manager

Not ready to move to netty yet, improve the existing network manager
---
 .../java/net/minecraft/server/DedicatedServer.java |  2 +-
 .../java/net/minecraft/server/NetworkManager.java  | 67 +++++++++++++---------
 .../net/minecraft/server/Packet56MapChunkBulk.java |  3 +-
 src/main/java/org/spigotmc/PacketQueue.java        | 51 ++++++++++++++++
 4 files changed, 93 insertions(+), 30 deletions(-)
 create mode 100644 src/main/java/org/spigotmc/PacketQueue.java

diff --git a/src/main/java/net/minecraft/server/DedicatedServer.java b/src/main/java/net/minecraft/server/DedicatedServer.java
index aefe20f..3d44192 100644
--- a/src/main/java/net/minecraft/server/DedicatedServer.java
+++ b/src/main/java/net/minecraft/server/DedicatedServer.java
@@ -94,7 +94,7 @@ public class DedicatedServer extends MinecraftServer implements IMinecraftServer
 
         try {
             // Spigot start
-            this.r = (!Boolean.getBoolean("org.spigotmc.netty.disabled"))
+            this.r = (!Boolean.getBoolean("org.spigotmc.netty.disabled")) && false // EMC - Disable netty
                     ? new org.spigotmc.netty.NettyServerConnection(this, inetaddress, this.G())
                     : new DedicatedServerConnection(this, inetaddress, this.G());
             // Spigot end
diff --git a/src/main/java/net/minecraft/server/NetworkManager.java b/src/main/java/net/minecraft/server/NetworkManager.java
index 7c64e9b..84a192a 100644
--- a/src/main/java/net/minecraft/server/NetworkManager.java
+++ b/src/main/java/net/minecraft/server/NetworkManager.java
@@ -1,5 +1,7 @@
 package net.minecraft.server;
 
+import org.spigotmc.PacketQueue; // EMC
+
 import java.io.BufferedOutputStream;
 import java.io.DataInputStream;
 import java.io.DataOutputStream;
@@ -8,15 +10,14 @@ import java.net.Socket;
 import java.net.SocketAddress;
 import java.net.SocketException;
 import java.security.PrivateKey;
-import java.util.ArrayList;
-import java.util.Collections;
-import java.util.Iterator;
-import java.util.List;
+import java.util.*;
 import java.util.concurrent.atomic.AtomicInteger;
 import javax.crypto.SecretKey;
 
 import java.io.IOException; // CraftBukkit
 
+
+
 public class NetworkManager implements INetworkManager {
 
     public static AtomicInteger a = new AtomicInteger();
@@ -29,8 +30,11 @@ public class NetworkManager implements INetworkManager {
     private volatile boolean m = true;
     private volatile boolean n = false;
     private java.util.Queue inboundQueue = new java.util.concurrent.ConcurrentLinkedQueue(); // CraftBukkit - Concurrent linked queue
-    private List highPriorityQueue = Collections.synchronizedList(new ArrayList());
-    private List lowPriorityQueue = Collections.synchronizedList(new ArrayList());
+    // EMC start
+    PacketQueue.PacketCounter counter = new PacketQueue.PacketCounter();
+    public PacketQueue highPriorityQueue = new PacketQueue(counter);
+    public PacketQueue lowPriorityQueue = new PacketQueue(counter);
+    // EMC end
     private Connection connection;
     private boolean s = false;
     private Thread t;
@@ -38,7 +42,7 @@ public class NetworkManager implements INetworkManager {
     private String v = "";
     private Object[] w;
     private int x = 0;
-    private int y = 0;
+    private volatile int y = 0; // EMC
     public static int[] c = new int[256];
     public static int[] d = new int[256];
     public int e = 0;
@@ -75,12 +79,11 @@ public class NetworkManager implements INetworkManager {
 
     public void queue(Packet packet) {
         if (!this.s) {
-            Object object = this.h;
-
-            synchronized (this.h) {
-                this.y += packet.a() + 1;
-                this.highPriorityQueue.add(packet);
-            }
+            // EMC start
+            this.y += packet.a() + 1;
+            List<Packet> queue = packet.lowPriority ? lowPriorityQueue : highPriorityQueue;
+            queue.add(packet);
+            // EMc end
         }
     }
 
@@ -89,10 +92,10 @@ public class NetworkManager implements INetworkManager {
 
         try {
             Packet packet;
-            int i;
-            int[] aint;
+            //int i;
+            //int[] aint;
 
-            if (this.e == 0 || !this.highPriorityQueue.isEmpty() && System.currentTimeMillis() - ((Packet) this.highPriorityQueue.get(0)).timestamp >= (long) this.e) {
+            if (!this.highPriorityQueue.isEmpty()) {
                 packet = this.a(false);
                 if (packet != null) {
                     Packet.a(packet, this.output);
@@ -104,22 +107,24 @@ public class NetworkManager implements INetworkManager {
                         this.k();
                     }
 
-                    aint = d;
-                    i = packet.k();
-                    aint[i] += packet.a() + 1;
+                    //aint = d;
+                    //i = packet.k();
+                    //aint[i] += packet.a() + 1;
                     flag = true;
                 }
             }
 
             // CraftBukkit - don't allow low priority packet to be sent unless it was placed in the queue before the first packet on the high priority queue TODO: is this still right?
-            if ((flag || this.lowPriorityQueueDelay-- <= 0) && !this.lowPriorityQueue.isEmpty() && (this.highPriorityQueue.isEmpty() || ((Packet) this.highPriorityQueue.get(0)).timestamp > ((Packet) this.lowPriorityQueue.get(0)).timestamp)) {
+            long lowId = this.lowPriorityQueue.orderId(); // EMC
+            long highId = this.highPriorityQueue.orderId(); // EMC
+            if (lowId != 0 && (highId == 0 || lowId < highId)) { // EMC
                 packet = this.a(true);
                 if (packet != null) {
                     Packet.a(packet, this.output);
-                    aint = d;
-                    i = packet.k();
-                    aint[i] += packet.a() + 1;
-                    this.lowPriorityQueueDelay = 0;
+                    //aint = d;
+                    //i = packet.k();
+                    //aint[i] += packet.a() + 1;
+                    //this.lowPriorityQueueDelay = 0;
                     flag = true;
                 }
             }
@@ -135,9 +140,14 @@ public class NetworkManager implements INetworkManager {
     }
 
     private Packet a(boolean flag) {
-        Packet packet = null;
-        List list = flag ? this.lowPriorityQueue : this.highPriorityQueue;
-        Object object = this.h;
+        PacketQueue list = flag ? this.lowPriorityQueue : this.highPriorityQueue;
+        Packet packet = list.poll();
+        if (packet != null) {
+            this.y -= packet.a() + 1;
+        }
+        return packet;
+        // EMC start
+        /*Object object = this.h;
 
         synchronized (this.h) {
             while (!list.isEmpty() && packet == null) {
@@ -149,7 +159,8 @@ public class NetworkManager implements INetworkManager {
             }
 
             return packet;
-        }
+        }*/
+        // EMC end
     }
 
     private boolean a(Packet packet, boolean flag) {
diff --git a/src/main/java/net/minecraft/server/Packet56MapChunkBulk.java b/src/main/java/net/minecraft/server/Packet56MapChunkBulk.java
index 8486d82..2cc3539 100644
--- a/src/main/java/net/minecraft/server/Packet56MapChunkBulk.java
+++ b/src/main/java/net/minecraft/server/Packet56MapChunkBulk.java
@@ -30,9 +30,10 @@ public class Packet56MapChunkBulk extends Packet {
     // CraftBukkit end
     private World world; // Spigot (Orebfuscator) Keep track of world
 
-    public Packet56MapChunkBulk() {}
+    public Packet56MapChunkBulk() {this.lowPriority=true;} // EMC - Make low priority to fix sign issues and match 51
 
     public Packet56MapChunkBulk(List list) {
+        this.lowPriority=true; // EMC - Make low priority to fix sign issues and match 51
         int i = list.size();
 
         this.c = new int[i];
diff --git a/src/main/java/org/spigotmc/PacketQueue.java b/src/main/java/org/spigotmc/PacketQueue.java
new file mode 100644
index 0000000..d29e81a
--- /dev/null
+++ b/src/main/java/org/spigotmc/PacketQueue.java
@@ -0,0 +1,51 @@
+package org.spigotmc;
+
+import net.minecraft.server.Packet;
+
+import java.util.AbstractList;
+import java.util.concurrent.ConcurrentLinkedQueue;
+
+public class PacketQueue extends AbstractList<Packet> {
+    ConcurrentLinkedQueue<Packet> queue = new ConcurrentLinkedQueue<Packet>();
+    ConcurrentLinkedQueue<Long> orderQueue = new ConcurrentLinkedQueue<Long>();
+
+    public static class PacketCounter {
+        long id = 1;
+        public long next() {
+            return id++;
+        }
+    }
+
+    PacketCounter counter;
+    public PacketQueue(PacketCounter counter) {
+        this.counter = counter;
+    }
+
+    @Override
+    public boolean add(Packet element) {
+        orderQueue.offer(counter.next());
+        return queue.offer(element);
+    }
+
+    @Override
+    public Packet get(int index) {
+        return poll();
+    }
+
+    public Packet poll() {
+        orderQueue.poll();
+        return queue.poll();
+    }
+    public Long orderId() {
+        Long res = orderQueue.peek();
+        return res == null ? 0 : res;
+    }
+    @Override
+    public int size() {
+        return queue.isEmpty() ? 0 : 1;
+    }
+    @Override
+    public boolean isEmpty() {
+        return queue.isEmpty();
+    }
+}
-- 
1.8.1.1

